File: week1/credit/credit.c
// Checks if credit card number is valid

#include <cs50.h>
#include <stdio.h>

long power(int base, int power);
int get_digit(long number, int place);
int get_number_length(long number);
bool check_sum(long number);

int main(void)
{
    // Get user input
    long number;
    do
    {
        number = get_long("Number: ");
    }
    while (number < 0);

    int numberLength = get_number_length(number);
    string answer = "INVALID";

    //  If valid card
    if (check_sum(number))
    {
        int firstTwo = 10 * get_digit(number, numberLength - 1) + get_digit(number, numberLength - 2);

        // Check for American Express card
        if (numberLength == 15 && (firstTwo == 34 || firstTwo == 37))
        {
            answer = "AMEX";
        }

        // Check for Master Card
        if (numberLength == 16 && (firstTwo > 50 && firstTwo < 56))
        {
            answer = "MASTERCARD";
        }

        // Check for Visa Card
        if ((numberLength == 13 || numberLength == 16) && (firstTwo / 10 == 4))
        {
            answer = "VISA";
        }

    }

    printf("%s\n", answer);
}

// Implements Luhn's algorithm and returns true/false
bool check_sum(long number)
{
    int numberLength = get_number_length(number);
    int sum = 0;

    // For every other digit starting from second to last
    for (int i = 1; i < numberLength; i = i + 2)
    {
        int n = get_digit(number, i) * 2;
        if (n > 9)
        {
            sum = sum + (n / 10) + (n % 10);
        }
        else
        {
            sum = sum + n;
        }
    }

    // Adds the rest of the numbers
    for (int i = 0; i < numberLength; i = i + 2)
    {
        sum = sum + get_digit(number, i);
    }


    return (sum % 10 == 0);
}

// Returns length of given number
int get_number_length(long number)
{
    int length = 0;

    while (true)
    {
        if (number / power(10, length) > 0)
        {
            length++;
        }
        else
        {
            break;
        }
    }

    return (length);
}

// Returns the given place of given number (starting with 0)
int get_digit(long number, int place)
{
    return ((number / power(10, place)) % 10);
}

// Returns base^power
long power(int base, int power)
{
    long answer = 1;
    for (int i = 0; i < power; i++)
    {
        answer = answer * base;
    }

    return answer;
}

File: sentimental-credit/credit.py
# Determines whether a credit card number is valid
# And if so, what type of card it is

def main():
    card_num = get_user_input()
    card = CreditCard(card_num)
    print(card.get_type())


# Gets a valid user input that consists only of numbers
def get_user_input():
    while True:
        try:
            number = int(input("Number: "))
            break
        except:
            pass
    return str(number)


# Credit Card object
class CreditCard:
    def __init__(self, number):
        self.number = number

    # Returns the type of credit card
    def get_type(self):
        if self.luhn_alg():
            n = self.number
            length = len(n)
            first_two = int(n[0:2])

            if length == 15 and (first_two == 34 or first_two == 37):
                return "AMEX\n"
            elif length == 16 and first_two > 50 and first_two < 56:
                return "MASTERCARD\n"
            elif (length == 13 or length == 16) and first_two // 10 == 4:
                return "VISA\n"
        return "INVALID\n"

    # Returns whether the credit card is a valid number using luhn's algorithm
    def luhn_alg(self):
        n = self.number
        sum = 0

        # For every other number beginning from second to last digit
        start1 = 0 if len(n) % 2 == 0 else 1
        for i in n[start1::2]:
            tmp = int(i)*2
            if tmp > 9:
                sum += (tmp % 10) + 1
            else:
                sum += tmp

        # Add remaining numbers
        start2 = 1 if len(n) % 2 == 0 else 0
        for j in n[start2::2]:
            sum += int(j)

        return sum % 10 == 0


main()
